hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java

```diff
@@ -119,8 +119,13 @@ public boolean isNewConfigurationNeeded() {
                 CSLog.d(TAG, "isNewConfigurationNeeded - Modem: " + mModem);
                 CSLog.d(TAG, "isNewConfigurationNeeded - Carrier Config Id: " + mConfigId);
             }
-            // Actual: (this.mConfigId == null && TextUtils.isEmpty(this.mModem)) ? false : true
-            return mConfigId != null || !TextUtils.isEmpty(mModem);
+            try {
+                return (mConfigId != null || !TextUtils.isEmpty(mModem)) && !ModemSwitcher.getCurrentModemConfig().equals(mModem);
+            } catch (IOException e) {
+                CSLog.w(TAG, "Unable to read out current configuration");
+            	// Actual: (this.mConfigId == null && TextUtils.isEmpty(this.mModem)) ? false : true
+				return mConfigId != null || !TextUtils.isEmpty(mModem);
+            }
```

Comment:
I think this part misses actually required updates:

- Note that if `isNewConfigurationNeeded` yields True then `set` will be called setting the config ID **and** modem.
- `mModem` is only non-Empty if it is not equal to `ModemSwitcher.getCurrentModemConfig()`, see `evaluateModem`
- So this means this condition can be rewritten/simplified to `(mConfigId != null && !ModemSwitcher.getCurrentModemConfig().equals(mModem)) || !TextUtils.isEmpty(mModem)) ;`

Given the logic in  `evaluateModem` already taking into account if any changes should be done I assume the same for `evaluateCarrierConfigId` and indeed it checks against `SystemProperties.get(PROP_TA_AC_VERSION, null)`. However that is never set although it should.

So indeed it updates (and reboots) even when it doesn't need to. That is actually a great find from you! I'll fix this together with some bits of other suspicious logic which I think is that way it is due to incomplete decompilation of the original source

Reply:
@antoniu200 Closing in on the IMS change and am wondering: Have you tested this part of the change and has it the effect you wanted? Because I can't see how. Thinking about this I believe the change here is a no-op:

You basically added `&& !ModemSwitcher.getCurrentModemConfig().equals(mModem)` to the current condition.

- `mModem` is set to the result of `evaluateModem`/`getModemConfigurationNeeded`
- The only codepath that returns a non-empty string is returning [`getModemConfiguration(variant)`](https://github.com/Flamefire/android_device_sony_yoshino-common/blob/d97d29f4e207026c8a0fe87b91f7684ae0291d11/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/ModemConfiguration.java#L55)
- However it only returns that if `modemConfiguration.equalsIgnoreCase(currentModemConfig)

Hence `mModem` can only be non-empty when it isn't equal to `ModemSwitcher.getCurrentModemConfig()`   
`ModemSwitcher.getCurrentModemConfig()` is never empty (unless an error occured) so an empty `mModem` is different.

Hence `!ModemSwitcher.getCurrentModemConfig().equals(mModem)` is always true and the condition has no effect.

Reply:
Yes, I have tested this. The change actually happened because, after I set `ir51_ir92` as default, I was prompted to reboot when swapping SIMs that have no particular modem configuration. The added check seemed to fix the issue.

I was assuming mModem was the new modem to be applied, at least that's what I understood from reading CustomizationSelector's logs.

But I will take a look on my PC in a minute, so I can read the code better.

Reply:
Oh, wow. I was staring at the reason why that added condition fixed the reboot issue for almost 20 minutes, trying to explain below why I don't understand what is going on...
When the whole reason was right under my eyes:

`isNewConfigurationNeeded()` originally returned at line 123 a boolean value based on a logic of the sorts:
Is a new CarrierConfig needed? If yes, we need to reboot.
Is a new Modem needed? If yes, we need to reboot.

So, basically, the reason my phone was rebooting is most likely because I still had the original CarrierConfigs, with one for DIGI added.
So, I was swapping two SIMs that needed no Modem update, but needed a CarrierConfig update.

This is probably why my Sony lost IMS services on DIGI after a while: after upgrading and downgrading between my own builds, the CarrierConfig probably got stuck on the Orange one, hence it only works on Orange now.

**That line of code is not needed anymore.**

Given that you have done the refactoring, should I update my PR, or just leave it as it is now?

~Okay, here are my 2 cents:~

~`Configurator.java`, line 115:~
~`mModem = evaluateModem(mBundle.getString(KEY_MODEM, ""));`~
~This represents the new modem to be applied, if needed, and executes `evaluateModem`.~

~`Configurator.java`, line 85:~
~`return new ModemConfiguration(getPreferences()).getModemConfigurationNeeded(modem);`~
~This executes `getModemConfigurationNeeded`.~

~`ModemConfiguration.java`, line 50:~
~`String currentModemConfig = ModemSwitcher.getCurrentModemConfig();`~
~This identifies `currentModemConfig`, exactly in the same fashion as I did above.~

~`ModemConfiguration.java`, line 52:~
~`if (ModemSwitcher.SINGLE_MODEM_FS.equalsIgnoreCase(currentModemConfig))`~
~It compares `currentModemConfig` to `SINGLE_MODEM_FS`, which is a string equal to `single_filesystem`.~
~When is this ever true?~

~`ModemConfiguration.java`, line 55:~
~`String modemConfiguration = getModemConfiguration(variant);`~
~`modemConfiguration` is the modem that needs to be applied, based on the `variant` parsed by `ModemConfParser.parseModemFileName()`.~

~`ModemConfiguration.java`, line 56:~
~`if (!TextUtils.isEmpty(modemConfiguration) && !modemConfiguration.equalsIgnoreCase(currentModemConfig))`~
~`modemConfiguration` is then compared to `currentModemConfig` as defined above.~

Reply:
> Given that you have done the refactoring, should I update my PR, or just leave it as it is now?

I moved the actual change into a suitable place ([as discussed](https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40#discussion_r1493824251)) with the additional check for IMS to be save, see [this](https://github.com/Flamefire/android_device_sony_yoshino-common/pull/40#discussion_r1494129810) and opened #42 mentioning your handle in the commit. So you don't need to do anything. If you care about commit authorship feel free to commit that yourself and open a PR.

I'll double check the other commits and will merge them soon-ish. https://github.com/antoniu200/android_packages_apps_Settings/commit/907d5897ed90bf1e210c6c42bd958eac388e6c5d looks fine, https://github.com/antoniu200/android_frameworks_opt_telephony/commit/959dcadfec2f7a44f8d884c28f9fa5321ddf6ee0 needs a small change:

```
        boolean provisioned = ImsManager.getInstance(mPhone.getContext(), mPhone.getPhoneId()).isVolteProvisionedOnDevice();
        if (!provisioned && isVolteEnabled) { // Not provisioned, but registered? Carrier probably needs no provisioning.
            ImsManager.getInstance(mPhone.getContext(), mPhone.getPhoneId()).setVolteProvisioned(true);
        }
```

-> 

```
        if (isVolteEnabled) {
            ImsManager imsManager = getImsManager();
            if (!imsManager.isVolteProvisionedOnDevice()) // Not provisioned, but registered? Carrier probably needs no provisioning.
                imsManager.setVolteProvisioned(true); 
        }
```

Basically:
- `ImsManager.getInstance` is used 6 times -> Add a new private function
- No need to check for provisioning if it isn't enabled
- Reuse the `ImsManager` used in the same function

Same for the other 2 changes. And maybe adjust the commit message when you amend&force-push that (would be easiest to create a patch out of it) as currently:
> Otherwise, we end up either enabling more features than available, either none.

", either" should have been likely simply "or". But with the actual change and its comment I'd guess the commit message in total could simply say: "Set provision status for carriers that do not need provisioning", also no need to mention the change file in the commit title. A prefix of e.g. "IMS: " might be better, i.e. the (sub-)system that is changed (like here the commits have a prefix like "OpenCS: " or "SEPolicy: " (if at all)

Reply:
I modified my above message a thousandth time, that part of the code drove me nuts: very poor variable naming.
The modification would be nice to have, please check the message again: it actually is very much possible to reboot for nothing.

As for the bug I had with my Sony last time, I fixed it. It was because my provisioning mod required the user to toggle the tech on inside Settings, but that toggle would not be available if the tech was not provisioned: vicious circle.

Hence, the solution was to enable the Setting by default.

> I'd guess the commit message in total could simply say: "Set provision status for carriers that do not need provisioning"

Yeah, you could do that, but I chose to write what I wrote as I wanted to explain exactly why I made the changes.

Thing is, Android allows us to disable provisioning when it's not available. Problem is, as mentioned before, we cannot make a million CarrierConfigs, we live in a world where everything should work properly automatically. Also, had I chosen to disable provisioning universally, carriers that needed provisioning would still not have IMS services.

And these features can work automatically very nicely, even when a provisioning server is missing: just check if the phone can register itself in the IMS server for that specific tech. If yes, it means the tech exists and provisioning for it can be set true.

Had I chosen to not do that modification, I risked not enabling IMS techs for carriers that do not need provisioning, as you say, or risked enabling all techs all the time, while some may be missing on the carrier, or on the SIM, etc, which would cause busy tones and calls not going through, hence my "end of the world" joke.

Reply:
> I modified my above message a thousandth time, that part of the code drove me nuts: very poor variable naming. The modification would be nice to have, please check the message again: it actually is very much possible to reboot for nothing.

I checked the message again and you seemingly added 

> So, I was swapping two SIMs that needed no Modem update, but needed a Customization ID update.
> [That added condition] actually still is needed, but can be reduced to TextUtils.isEmpty(mModem).

Which I don't understand. So you do need a Customization ID update but at the same time say you don't need to check for and do that?

To me that first sentence states that the condition should be left as it is now. With the refactoring done it should be a bit easier to understand

Reply:
Understandable, let me rephrase.

The condition I added fixed a reboot issue I had that was caused by me adding a DIGI Romania modem customization that was identical to the Orange Romania one.

Currently, `isNewConfigurationNeeded()` returns true if a customization change is needed (regardless of modem) **or** if a modem change is needed.

So, basically, we can currently reboot just because a line residing in file inside `/data/data` (forgot the exact path of OpenCS) was updated.

This update can (and most likely does) happen without rebooting, only the modem update requires a reboot.

Hence, that return value still needs a change, and the best change is to only check if `mModem` is not empty.

Reply:
> Currently, `isNewConfigurationNeeded()` returns true if a customization change is needed (regardless of modem) **or** if a modem change is needed.
> 
> So, basically, we can currently reboot just because a line residing in file inside `/data/data` (forgot the exact path of OpenCS) was updated.

But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.

From what I understand what happens is: You write to MiscTA. Upon reboot the hardware/driver will write into a property. This property is then checked against the desired value in `evaluateConfig` which will return `null` if no change is required which makes this function return False, hence no update and no reboot will happen.   
However I haven't found what/where exactly that property is written, so I just assume it is on reboot.

I fixed a potential issue where a change to that config was detected even if it isn't required: https://github.com/Flamefire/android_device_sony_yoshino-common/commit/342abc1c9e22650ae5d3b5690f843a0c4dbef843

I also improved the logging to provide the value in the property and the MiscTA to aid in debugging: https://github.com/Flamefire/android_device_sony_yoshino-common/blob/dc325bcf9b5d720e0a8fc8ced82b8126ded6b597/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java#L84-L86

> This update can (and most likely does) happen without rebooting, only the modem update requires a reboot.

As mentioned I'm not sure about that so I'd rather reboot one more time than one to less. I mean it only happens when changing SIMs, doesn't it? That's rare enough. Furthermore at this point you are actually **preventing** the change in the first place, so with only checking the modem here the config will **not** be updated if it would only be the config in need of updating. This is much more likely after defaulting to the ir-modem for IMS.

> **EDIT:** My God, all these changes and comments I made yesterday were probably not even visible to you, because they were "Pending". Great.
> 
> Well, I hope you can see them now.

You mean the ones on `strings.xml`? If so yes.

Reply:
> But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.

Oh, no, it is not the case. We are not changing CarrierConfigs using OpenCS, but rather only modems, which are specified using Customization IDs (`mConfigID` in the code above).

> From what I understand what happens is: You write to MiscTA.

I have to check this, but, AFAIK, MiscTA is in /oem somewhere, however my /oem is completely empty and OpenCS works fine.

> I fixed a potential issue where a change to that config was detected even if it isn't required: [342abc1](https://github.com/Flamefire/android_device_sony_yoshino-common/commit/342abc1c9e22650ae5d3b5690f843a0c4dbef843)
> 
> I also improved the logging to provide the value in the property and the MiscTA to aid in debugging:
> 
> https://github.com/Flamefire/android_device_sony_yoshino-common/blob/dc325bcf9b5d720e0a8fc8ced82b8126ded6b597/hardware/OpenCustomizationSelector/src/com/sonymobile/customizationselector/Configurator.java#L84-L86

Will check these a bit later on my PC as well.

> Furthermore at this point you are actually **preventing** the change in the first place, so with only checking the modem here the config will **not** be updated if it would only be the config in need of updating. This is much more likely after defaulting to the ir-modem for IMS.

The point is definitely not to prevent the change, but rather just the reboot.

However, I am not sure I understand what you mean by "config" here. OpenCS' config or MiscTA? Since the first one is definitely updated at runtime.

> > **EDIT:** My God, all these changes and comments I made yesterday were probably not even visible to you, because they were "Pending". Great.
> > Well, I hope you can see them now.
> 
> You mean the ones on `strings.xml`? If so yes.

Good. Even my previous comment was pending for a while before I realised what I had to do to publish it.

Reply:
> > But we need to update the customization (i.e. MiscTA) if a customization change is needed. If only checking for a modem change than we will miss that change which likely will lead to missing features.
> 
> Oh, no, it is not the case. We are not changing CarrierConfigs using OpenCS, but rather only modems, which are specified using Customization IDs (`mConfigID` in the code above).

There is `mConfigID` and `mModem`. Both are written to different parts in the `MiscTA` "partition" using the class of the same name.

> The point is definitely not to prevent the change, but rather just the reboot.

Then this approach does not work as this function is used by `set` to determine if anything needs setting and by callers to determine if `set` needs to be called.

> However, I am not sure I understand what you mean by "config" here. OpenCS' config or MiscTA? Since the first one is definitely updated at runtime.

`mConfig` basically.

Reply:
What I notice is that writing to MiscTA is done using a function called `writeRAW()`, but I cannot find it anywhere in the repo. So, does that function work? If yes, where is it?

Reply:
`writeRaw` is declared as "native", hence it lives inside a native binary, not in Java. In this case you can even see which one easily as it is mentioned right next to it: https://github.com/Flamefire/android_device_sony_yoshino-common/blob/9202c964dd848cf507530f4e7483eb74495ea1ce/hardware/OpenCustomizationSelector/src-misc-ta/com/sonymobile/miscta/MiscTA.java#L14

See https://github.com/Flamefire/android_device_sony_lilac/blob/1695c51519186d66928439ca5ed65c7263380866/proprietary-files.txt#L93

Reply:
Okay, I do not like the current solution, but it seems we have to leave it like this, as of now. I cannot tell what `writeRAW` from inside `libMiscTAApi.sony.so` does, as I barely have any experience with ARMv7 assembly and, besides, ARM decompilers can barely interpret the results of C code intended for hardware / drivers: many missing external functions, function calls, wrong and unfixable function arguments, etc.